src/hotspot/share/gc/parallel/psParallelCompact.cpp

```diff
@@ -1722,36 +1729,165 @@ class PSAdjustTask final : public WorkerTask {
+  assert(num_workers > 0, "precondition");
+  assert(worker_id < num_workers, "precondition");
+
+  size_t num_regions = end - start;
+  size_t num_regions_per_worker = num_regions / num_workers;
+  size_t remainder = num_regions % num_workers;
+  // The first few workers will get one extra.
+  *worker_start = start + worker_id * num_regions_per_worker
+                  + MIN2(checked_cast<size_t>(worker_id), remainder);
+  *worker_end = *worker_start + num_regions_per_worker
+                + (worker_id < remainder ? 1 : 0);
+}
+
+void PSParallelCompact::forward_to_new_addr() {
+  GCTraceTime(Info, gc, phases) tm("Forward", &_gc_timer);
+  uint nworkers = ParallelScavengeHeap::heap()->workers().active_workers();
+
+  struct ForwardTask final : public WorkerTask {
+    uint _num_workers;
+
+    explicit ForwardTask(uint num_workers) :
+      WorkerTask("PSForward task"),
+      _num_workers(num_workers) {}
+
+    void work(uint worker_id) override {
+      ParCompactionManager* cm = ParCompactionManager::gc_thread_compaction_manager(worker_id);
+      for (uint id = old_space_id; id < last_space_id; ++id) {
+        MutableSpace* sp = PSParallelCompact::space(SpaceId(id));
+        HeapWord* dense_prefix_addr = dense_prefix(SpaceId(id));
+        HeapWord* top = sp->top();
+
+        if (dense_prefix_addr == top) {
+          continue;
+        }
+
+        size_t dense_prefix_region = _summary_data.addr_to_region_idx(dense_prefix_addr);
+        size_t top_region = _summary_data.addr_to_region_idx(_summary_data.region_align_up(top));
+        size_t start_region;
+        size_t end_region;
+        split_regions_for_worker(dense_prefix_region, top_region,
+                                 worker_id, _num_workers,
+                                 &start_region, &end_region);
+        for (size_t cur_region = start_region; cur_region < end_region; ++cur_region) {
+          RegionData* region_ptr = _summary_data.region(cur_region);
+          size_t live_words = region_ptr->partial_obj_size();
+
+          if (live_words == ParallelCompactData::RegionSize) {
+            // No obj-start
+            continue;
+          }
+
+          HeapWord* region_start = _summary_data.region_to_addr(cur_region);
+          HeapWord* region_end = region_start + ParallelCompactData::RegionSize;
+
+          HeapWord* cur_addr = region_start + live_words;
+
+          HeapWord* destination = region_ptr->destination();
+          while (cur_addr < region_end) {
+            cur_addr = mark_bitmap()->find_obj_beg(cur_addr, region_end);
+            if (cur_addr >= region_end) {
+              break;
+            }
+            assert(mark_bitmap()->is_marked(cur_addr), "inv");
+            HeapWord* new_addr = destination + live_words;
+            oop obj = cast_to_oop(cur_addr);
+            if (new_addr != cur_addr) {
+              cm->preserved_marks()->push_if_necessary(obj, obj->mark());
+              obj->forward_to(cast_to_oop(new_addr));
```

Comment:
>> The live_words is firstly initialized as partial_obj_size of the current region
>
> For non-spliting regions, destination is the destination for the first live word in the current region.

I know `destination` is the first live word of  the current region. But the new address seems to should be `destination + <live words>` instead of `destination + <live words> + <partial_obj_size>`.

Reply:
> the new address seems to should be destination + <live words>

Isn't that `new_addr = destination + live_words`?

(Not super clear what you are suggesting; can you post some code snippet here or point me to your repo/branch with your suggestion so that I can see the exact code?)

Reply:
> > the new address seems to should be destination + `<live words>`
> 
> Isn't that `new_addr = destination + live_words`?

```
size_t live_words = region_ptr->partial_obj_size();
```

But the variable `live_words` is initialized as `partial_obj_size` (shown above). If the `partial_obj_size` is not zero, the code `new_addr = destination + live_words` actually equals to `new_addr = destination + <the actual live words> + <partial_obj_size>`. Is this your intention?

Reply:
`partial_obj_size` is non-zero iff those words belongs to a live obj. If you change to `size_t live_words = 0;`, wouldn't you get a test-failure?

Reply:
I know what I missed now. The method  `RegionData::data_size`, used by `ParallelCompactData::summarize`, includes the `partial_obj_size` and `live_obj_size`. Then the `destination` is a previse location of a region (split by the the region border instead of the object border)